src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -254,7 +260,9 @@ bool ClusterTaskManager::CancelTask(const TaskID &task_id) {
        shapes_it++) {
     auto &work_queue = shapes_it->second;
     for (auto work_it = work_queue.begin(); work_it != work_queue.end(); work_it++) {
-      if (std::get<0>(*work_it).GetTaskSpecification().TaskId() == task_id) {
+      const auto &task = std::get<0>(*work_it);
+      if (task.GetTaskSpecification().TaskId() == task_id) {
+        RemoveFromBacklog(task);
```

Comment:
Actually do we need the backlog tracker? Why not instead calculate the max backlog size during the Heartbeat call, since you are looping over the tasks anyways?

Reply:
So we can do that, it just means to heartbeat will take O(n) time instead of O(k). n=# of tasks, k=# of scheduling classes

Reply:
Yes, let's do that. It's actually just O(k) still since queueing happens on the core worker anyways.